File: 2-add-two-numbers/2-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* dummy= new ListNode();
        auto temp = new ListNode();
        temp=dummy;
        int carry = 0;
        while(l1||l2){
            auto tot = carry;
            if(l1){
                tot+=l1->val;
                l1 = l1->next;
            }
            
            if(l2){
                tot+=l2->val;
                l2 = l2->next;
            }
            
            if(tot>=10){
                carry = 1;
            }
            else{
                carry = 0;
            }
            temp->next = new ListNode(tot%10);
            temp = temp->next;
        }
        
        if(carry){
            temp->next = new ListNode(1);
        }
        return dummy->next;
    }
};

File: 2-add-two-numbers/2-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        dummy = ListNode()
        temp = dummy
        p1,p2,carry = l1,l2,0
        while p1 or p2:
            cur1,cur2 = p1.val if p1 else 0,p2.val if p2 else 0
            tot = cur1+cur2+carry
            curVal = tot%10
            node = ListNode(curVal)
            temp.next = node
            temp = temp.next
            if tot>=10:
                carry = 1
            else:
                carry = 0
            if p1: p1=p1.next
            if p2: p2=p2.next
        if carry:
            temp.next = ListNode(carry)
                
        return dummy.next